src/c++11/impl.cc

```diff
@@ -81,6 +103,71 @@ std::unique_ptr<SpanImpl> TracerImpl::StartSpan(std::shared_ptr<TracerImpl> self
   return span;
 }
 
+bool TracerImpl::inject(SpanContext sc, const CarrierFormat& format, const CarrierWriter& opaque) {
+  const BuiltinCarrierFormat *bcf = dynamic_cast<const BuiltinCarrierFormat*>(&format);
+  if (bcf == nullptr) {
+    return false;
+  }
+  switch (bcf->type()) {
+  case BuiltinCarrierFormat::HTTPHeaders:
+  case BuiltinCarrierFormat::TextMap:
+    break;
+  default:
+    return false;
+  }
+
+  const TextMapWriter* carrier = dynamic_cast<const TextMapWriter*>(&opaque);
+  if (carrier == nullptr) {
+    return false;
+  }
+  carrier->Set(FieldNameTraceID, uint64ToHex(sc.trace_id()));
+  carrier->Set(FieldNameSpanID, uint64ToHex(sc.span_id()));
+  carrier->Set(FieldNameSampled, sc.sampled() ? "true" : "false");
+
+  sc.ForeachBaggageItem([carrier](const std::string& key,
+				  const std::string& value) {
+			  carrier->Set(std::string(PrefixBaggage) + key, value);
+			  return false;
+			});
+  return true;
+}
+
+SpanContext TracerImpl::extract(const CarrierFormat& format, const CarrierReader& opaque) {
+  const BuiltinCarrierFormat *bcf = dynamic_cast<const BuiltinCarrierFormat*>(&format);
+  if (bcf == nullptr) {
+    // TODO error handling
+    return SpanContext();
+  }
+  switch (bcf->type()) {
+  case BuiltinCarrierFormat::HTTPHeaders:
+  case BuiltinCarrierFormat::TextMap:
+    break;
+  default:
+    return SpanContext();
+  }
+
+  const TextMapReader* carrier = dynamic_cast<const TextMapReader*>(&opaque);
+  if (carrier == nullptr) {
+    return SpanContext();
+  }
+
+  std::shared_ptr<ContextImpl> ctx(new ContextImpl);
+  carrier->ForeachKey([carrier, &ctx](const std::string& key,
+				const std::string& value) {
+			if (key == FieldNameTraceID) {
```

Comment:
In other language impls, we keep track of the required fields (i.e., trace_id, span_id, sampled) we find in the map and return null (or `SpanContext()` depending on how you want to think about my previous comment) when any required fields are missing. Pretty easy to do just by incrementing a counter and checking `==3` before returning the ctx.